torchmultimodal/modules/encoders/text_encoder.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Callable, Optional
+
+import torch
+from torch import nn, Tensor
+from torchmultimodal.modules.layers.text_embedding import BERTTextEmbeddings
+from torchmultimodal.modules.layers.transformer import (
+    TransformerEncoder,
+    TransformerOutput,
+)
+from torchmultimodal.utils.attention import get_extended_attention_mask
+
+
+class TextEncoder(nn.Module):
+    """
+    General text transformer encoder with embeddings, similar to BERT / RoBERTa.
+    Can be constructed with any user-provided embeddings and encoder.
+
+    Based on https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L870
+
+    Attributes:
+        embeddings (nn.Module): Module that projects text token ids into embeddings.
+            ``forward()`` should follow interface:
+                input_ids: Optional[Tensor],
+                token_type_ids: Optional[Tensor],
+                position_ids: Optional[Tensor],
+                inputs_embeds: Optional[Tensor],
+        encoder (nn.Module): Module for transformer encoder. ``forward()`` should follow interface:
+            Inputs:
+                hidden_states: Tensor,
+                attention_mask: Optional[Tensor], shape [batch_size, num_heads, query_seq_length, key_seq_length]
+            Returns:
+                ``TransformerOutput``
+        layernorm (nn.Module, optional): Module for layernorm to be applied after encoder, if provided.
+        pooler (nn.Module, optional): Module for pooler to be applied after layernorm, if provided.
+
+    Args:
+        input_ids (Tensor, optional): Tensor of input vocab token ids of shape [batch, seq_len].
+        attention_mask (Tensor, optional): Tensor indicating which tokens to attend to, shape [batch, seq_len]
+        token_type_ids (Tensor, optional): Tensor of input token type ids of shape [batch, seq_len]. In BERT,
+            used to indicate whether a word is in sentence A or B for next sentence prediction
+        position_ids (Tensor, optional): Tensor of input position ids of shape [batch, seq_len]
+        inputs_embeds (Tensor, optional): Tensor of input embeddings of shape [batch, hidden_size],
+            if embeddings are calculated elsewhere
+
+    Raises:
+        ValueError: if input_ids and inputs_embeds are both None
+    """
+
+    def __init__(
+        self,
+        embeddings: nn.Module,
+        encoder: nn.Module,
+        layernorm: Optional[nn.Module] = None,
+        pooler: Optional[nn.Module] = None,
+        weight_init_fn: Optional[Callable] = None,
+    ) -> None:
+        super().__init__()
+
+        self.embeddings = embeddings
+        self.encoder = encoder
+        self.layernorm = layernorm
```

Comment:
this is in addition to the one after the final layer of the transformer encoder?

Reply:
yes

Reply:
this should probably get upstreamed to transformer encoder? maybe add a comment